jobs/base_jobs.py

```diff
@@ -0,0 +1,161 @@
+    - Pipelines
+    - PTransforms
+    - PValues/PCollections
+    - Runners
+
+Pipelines manage a DAG (directed acyclic graph) of PValues and the PTransforms
+that compute them. Conceptually, PValues are the DAG's nodes and PTransforms are
+the edges.
+
+For example:
+
+    .------------. io.ReadFromText(fname) .-------. FlatMap(str.split)
+    | Input File | ---------------------> | Lines | -----------------.
+    '------------'                        '-------'                  |
+                                                                     |
+       .----------------. combiners.Count.PerElement() .-------.     |
+    .- | (word, count)s | <--------------------------- | Words | <---'
+    |  '----------------'                              '-------'
+    |
+    | MapTuple(lambda word, count: '%s: %d' % (word, count)) .------------.
+    '------------------------------------------------------> | "word: #"s |
+                                                             '------------'
+                                                                    |
+                             .-------------. io.WriteToText(ofname) |
+                             | Output File | <----------------------'
+                             '-------------'
+
+PCollections are PValues that represent a dataset of (virtually) any size,
+including unbounded/continuous datasets. They are the primary input and output
+types used by PTransforms.
+
+Runners provide a run() method for visiting every node (PValue) in the
+pipeline's DAG by executing the edges (PTransforms) to compute their values. At
+Oppia, we use DataflowRunner() to have our Pipelines run on the Google Cloud
+Dataflow service: https://cloud.google.com/dataflow.
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from jobs import job_options
+import python_utils
+
+
+class JobMetaclass(type):
+    """Metaclass for all of Oppia's Apache Beam jobs.
+
+    This class keeps track of the complete list of jobs. The list can be read
+    with the JobMetaclass.get_all_jobs() class method.
+
+    THIS CLASS IS AN IMPLEMENTATION DETAIL, DO NOT USE IT DIRECTLY. All user
+    code should simply inherit from the JobBase class, found below.
+    """
+
+    _JOB_REGISTRY = {}
+
+    def __new__(mcs, name, bases, namespace):
+        """Creates a new job class with type `JobMetaclass`.
+
+        https://docs.python.org/3/reference/datamodel.html#customizing-class-creation
+
+        This metaclass adds jobs to the _JOB_REGISTRY dict, keyed by name, as
+        they are created. We use the registry to reject jobs with duplicate
+        names and to provide the convenient: JobMetaclass.get_all_jobs().
+
+        We use a metaclass instead of other alternatives (like decorators or a
+        manual list), because metaclasses cannot be forgotten to be used,
```

Comment:
Didn't we use a manual list in job_registry? I thought it worked fine (especially once we added the lint checks).

That said I'm ok with trying this out and seeing how it goes. It's nice that it's an implementation detail, that means its impact can be limited (especially if we also make it private to this module).

However, I would suggest having a test that verifies that the actual output of get_all_jobs() is nonempty, to make sure that jobs are actually being picked up. We've had cases before where an oversight meant that 0 things were found.